R/utils.R

```diff
@@ -48,38 +62,94 @@ c_dedupl <- function(...) {
   }
 }
 
+reorder_explanatory <- function(x, order) {
+  x[[explanatory_name(x)]] <- factor(
+    explanatory_variable(x),
+    levels = c(order[1], order[2])
+  )
+  x
+}
+
+# Getters, setters, and indicators ------------------------------------------
+explanatory_expr <- function(x) {
+  attr(x, "explanatory")
+}
+
+explanatory_name <- function(x) {
+  as.character(explanatory_expr(x))
+}
+
 explanatory_variable <- function(x) {
-  x[[as.character(attr(x, "explanatory"))]]
+  if (!is.null(explanatory_expr(x))) {
+    x[[explanatory_name(x)]]
+  } else {
+    NULL
+  }
+}
+
+response_expr <- function(x) {
+  attr(x, "response")
+}
+
+response_name <- function(x) {
+  as.character(response_expr(x))
 }
 
-# Other places in the code use
-# dplyr::pull(x, !!attr(x, "response"))
-# which seems to do the same thing
 response_variable <- function(x) {
-  x[[as.character(attr(x, "response"))]]
+  x[[response_name(x)]]
 }
 
-reorder_explanatory <- function(x, order) {
-  x[[as.character(attr(x, "explanatory"))]] <- factor(
-    explanatory_variable(x),
-    levels = c(order[1], order[2])
+theory_type <- function(x) {
+  attr(x, "theory_type")
+}
+
+get_response_levels <- function(x) {
+  as.character(unique(response_variable(x)))
+}
+
+get_success_then_response_levels <- function(x) {
+  success_attr <- attr(x, "success")
+  response_levels <- setdiff(
+    get_response_levels(x),
+    success_attr
   )
-  x
+  c(success_attr, response_levels)
+}
+
+has_unused_levels <- function(x) {
+  present_levels <- unique(as.character(x))
+  unused_levels <- setdiff(levels(x), present_levels)
+  
+  length(unused_levels) > 0
+}
+
+is_generated <- function(x) {
+  attr(x, "generated")
+}
+
+is_hypothesized <- function(x){
+  attr(x, "hypothesized")
+}
+
+attr_is_null <- function(x, at) {
```

Comment:
Ultimately, I think we want to read these indicator functions like tests rather than questions. e.g. I read ` if (is.null(x))` like “if x is null” rather than “if is x null”. The object of the other indicator functions, `is_hypothesized` for example, is the first argument, so those cases are more straightforward. We have the inclination to specify the object in `is_nuat` because the component in question here is an attribute of the first argument rather than the argument itself, i.e. `is_null` doesn’t express what we’re trying to do here, so we need to include the object in the function name. The current `attr_is_null` phrasing here reads more like a natural language test (”if attribute is null”) than `is_attr_null` (“if is attribute null”) to me, but I’m fine to change this if you have strong feelings.

Reply:
After some digging, didn't find any "r specific" recommendation about naming "boolean" function. Many seem to use `is_*()` approach if it has single argument. Cases with multiple arguments are trickier to find, but here is one: [{dplyr}'s `attr_equal()`](https://github.com/tidyverse/dplyr/blob/0c9900a11854bdb49a376d3bb56948c39e00e1e7/R/utils.r#L75).

For other programming languages there is often the following recommendation on naming functions which return single boolean: it should take a form of a yes/no question/predicate. So that got me thinking: how about using an inverse `has_attr()`? This will lead to updates in many places, but seems to be the most precise alternative.

Reply:
After some digging, didn't find any "r specific" recommendation about naming "boolean" function. Many seem to use `is_*()` approach if it has single argument. Cases with multiple arguments are trickier to find, but here is one: [{dplyr}'s `attr_equal()`](https://github.com/tidyverse/dplyr/blob/0c9900a11854bdb49a376d3bb56948c39e00e1e7/R/utils.r#L75).

For other programming languages there is often the following recommendation on naming functions which return single boolean: it should take a form of a yes/no question/predicate. So that got me thinking: how about using an inverse `has_attr()`? This will lead to updates in many places, but seems to be the most precise alternative.

Reply:
I like `has_attr()`! Making the switch now.🙂